README
smoothscan 0.1.0

Description
-----------

smoothscan is a tool to convert scanned text into a vectorized output
form. Because printed text is assembled from fonts, each particular
letter (like 'o') will have the same shape as every other 'o' in the
document. We can take advantage of this, by building a table of such
symbols, and represent each occurrence of a symbol with a reference to
that symbol's table entry. This will save a lot of space, and a
similar idea is used in djvu's jb2 mode and JBIG2 for PDF.

smoothscan builds up this table, but instead of filling the table with
the original raster images, it vectorizes each symbol. Vector images
will look smoother than their raster equivalents, and can be scaled
without introducing pixelation. These properties result in a smaller
output file size, as well as making the scanned text images more
readable.

smoothscan saves the vectorized images into a custom TrueType font and
embeds the font into the output pdf file. Currently each symbol is
mapped to an arbitrary letter in the font, but in future versions you
could run OCR on each symbol, and ensure that the 'o' image is
associate with the 'o' character encoding in the generated font.

To get good results, you must have good input. Higher resolution scans
capture more detail about the shape of each symbol, so a higher
quality vectorized version can be created. It's a good idea to process
your scanned images using a tool like ScanTailor before running
smoothscan.

Current smoothscan can only process pure black and white 1bpp images,
but in the future support will be added for other formats, especially
ScanTailor's Mixed output mode.

smoothscan is currently targeted at GNU/Linux based systems, but
Windows and OS X will be supported in future versions.

Downloading
-----------

You can always find the latest release on the project home page at
<https://natecraun.net/projects/smoothscan/>. Developers can check out
the latest bleeding edge source code on the project's Github page:
<https://github.com/ncraun/smoothscan>

Installation
------------

Please see the INSTALL file for installation instructions.

News
----

Please see the NEWS file for a high level overview of changes since
the last release.

License
-------

smoothscan is licensed under the GPLv3+ Please see the COPYING file
for the full text of this license.

Authors
-------

Please see the AUTHORS file for the list of contributors.


NEWS
smoothscan 0.1.0: first release


src/smoothscan.h
/*
  This file is part of smoothscan.

  smoothscan is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  smoothscan is distributed in the hope that it will be useful, but
  WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
  General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with smoothscan. If not, see <http://www.gnu.org/licenses/>.
*/

#ifndef SMOOTHSCAN_H_INCLUDED
#define SMOOTHSCAN_H_INCLUDED

/* Change this for new releases */
#define SMOOTHSCAN_VERSION "0.1.0"

/* Represent the mapping from a symbol to a font code point */
struct mapping
{
  l_int32 iclass;		/* The symbol's index */
  unsigned char code_point;	/* The font code point */
  unsigned int font_num;	/* which font it belongs to */
  int used;			/* 1 if used, 0 if empty */
};

/* Hold the command line arguments for the program */
struct args
{
  /* Parameters */
  int num_input_files;
  char **input_files;
  char *outname;

  /* Optional Parameters */
  double thresh;
  double weight;

  /* Flags */
  int help_flag;
  int version_flag;

  /* Debug params */
  char *debug_tmpdir;
  /* Debug Flags */
  int debug_draw_borders;
  int debug_render_pages;
  int debug_skip_font_gen;
  int debug_no_clean_tmpdir;
};

/*
  Font segmentation:

  There is a limited number of codepoints available in the font, and
  we will likely need more than one font to include all the
  characters.

  Encoding used is KOI8-R, because of all the (non unicode) encodings
  supported by libharu and fontforge, KOI8-R allows for the most
  printable codepoints.

  With KOI8-R, we can encode from [33 (exclamation point) to 126
  (tilde)], [128 to 153], and [155 to 255]

  We have a total of 221 useable codepoints in KOI8-R
*/

/*
  Return the first valid code point of the font.
*/
unsigned char first_code_point ();

/*
  Return the last valid code point of the font.
 */
unsigned char max_code_point ();

/* 
   Return the next font code point after prev.
   
   next_code_point will NOT wrap around, you have to check against
   max_code_point yourself. However it will skip the nonprintable
   chars.
*/
unsigned char next_code_point (unsigned char prev);

/* 
   Print the text that is displayed when the user passes --help, and
   exit.
 */
void print_help ();

/*
  Print the text that is displayed when the user passes --version, and
  exit.
 */
void print_version ();

/*
  Prints str to stderr, and terminates the program.
 */
void error_quit (const char *str);

/*
  Return the number of digits that would be printed if a was
  represented in decimal form. Not valid for negative numbers.
 */
int num_digits (int a);

/*
  Callback for nftw to delete a file or a directory.
 */
int
delete_file (const char *path, const struct stat *sb, int typeflag,
	     struct FTW *ftwbuf);

/*
  Error handler callback for libharu. Simply prints libharu's
  generated error, and then error_quits itself.
 */
void
pdf_error_handler (HPDF_STATUS error_no, HPDF_STATUS detail_no,
		   void *user_data);

/*
  Out of memory errors should be considered as fatal errors for this
  program. This function is a wrapper around malloc that will check
  the malloc call for problems, and close the program if there is an
  out of memory error. Use this call instead of plain malloc.
*/
void *malloc_guarded (size_t size);

/*
  See if filename exists, and if we can read from it.

  Return 1 if it exists, 0 if it doesn't.
*/
int file_exists (const char *filename);

/*
  Create a font from the temp font directory with all the glyph image
  in it. Currently this just invoke the python font generation. 

  dirname - The directory the font is stored in.

  fontname - The output filename for the font, including the .ttf
  suffix

  latticeh, latticew - values from JBDATA

  fontnum - The internal number of the font (from the for loop).
 */
void
create_font_from_dir (const char *dirname, const char *fontname, int latticeh,
		      int latticew, int fontnum);

/*
  Generate the fonts that will be embedded in the output pdf.

  data - The JBDATA from leptonica's classifier.

  maps - Mapping each symbol to a font code point.

  num_fonts - The number of fonts to generate (calculated from the
  mapping phase)

  dir - The alternate directory to store files in. If NULL, then
  generate_fonts will use the value stored in the environmental var
  TMPDIR, if TMPDIR is empty it will use the value from POSIX's
  P_tmpdir, which should be something like /tmp or /var/tmp depending
  on your system.
 */
char *generate_fonts (const JBDATA * data, const struct mapping *maps,
		      int num_fonts, char *dir);

/*
  Create the pdf using libharu.

  outname - The filename to store the pdf into.

  tmpdirname - The temp directory where all the fonts are stored.

  num_fonts - The number of fonts.

  num_input_files - The number of input files (the number of pages in
  the pdf)

  data - JBDATA from leptonica
  
  maps - mappings from each symbol to its font code point

  debug_draw_borders - if 1, draw red rectangles where each glyph
  should be placed, if 0 don't.

*/
void
generate_pdf (const char *outname, const char *tmpdirname, int num_fonts,
	      int num_input_files, const JBDATA * data,
	      const struct mapping *maps, int debug_draw_borders);

/*
  Use leptonica to create the JBDATA, which is the dictionary of all
  the different symbols in the document.

  input_files - The list of 1bpp images (1 image per page).
  
  thresh - Specify the threshold value (value for correlation). Valid
  input is from [0.40 - 0.98]. Recommended values for scanned text
  from [0.80 - 0.85].  Default is 0.85.

  weight - Specify the weight value (correcting threshold for thick
  characters).  Valid input is from [0.0 - 1.0].  Recommended values
  for scanned text from [0.5 - 0.6].  Default is 0.5.
*/
JBDATA *classify_components (int num_input_files, char **input_files,
			     double thresh, double weight);


/*
  Map each symbol to a code point in the font.

  Returns the number of fonts needed.

  data - The leptonica JBDATA dictionary.

  in_maps - This is actually an output variable, it will be modified
  to hold all the generate mappings. It will be allocated in
  register_mappings, so it's up to the caller to free it.
*/
int register_mappings (const JBDATA * data, struct mapping **in_maps);

/*
  Create an arg struct with default parameters, and change them from
  the default according to the command line args. Uses
  getopt_long. Pass it the argc and argv from main.
*/
struct args *parse_args (int argc, char *argv[]);

/*
  Make sure all the command line arguments are valid.

  Return 1 if valid, error_quit if invalid.
*/
int validate_args (const struct args *args);

#endif /* SMOOTHSCAN_H_INCLUDED */


src/smoothscan-fontgen.py
#!/usr/bin/python

#  This file is part of smoothscan.
#
#  smoothscan is free software: you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published
#  by the Free Software Foundation, either version 3 of the License,
#  or (at your option) any later version.
#
#  smoothscan is distributed in the hope that it will be useful, but
#  WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
#  General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with smoothscan. If not, see <http://www.gnu.org/licenses/>.


# Fontforge has a c library interface, but there are no docs for it,
# and even the author recommends using the python interface instead of
# the C library interface.

import fontforge
import psMat
import glob
import os
import sys
import subprocess

def removePrefix(text, prefix):
    return text[len(prefix):] if text.startswith(prefix) else text

def removePostfix(text, postfix):
    return text[:len(text)-len(postfix)]

ffVersion = fontforge.version()
print ("Using Fontforge version: " + ffVersion)

if (len(sys.argv) != 6):
    print ("Usage: fontgen.py fontdir outname latticeh latticew fontnum")
    exit (1)

fontdir = sys.argv[1]
outname = sys.argv[2]
latticeh = int(sys.argv[3])
latticew = int(sys.argv[4])
fontnum = int(sys.argv[5])
# command line args

print ("Scaling to x: " + str(latticeh) + " y: " + str(latticeh))
print ("Converting " + fontdir  + "/*.png to " + outname)


newFont = fontforge.font() 
newFont.encoding = "koi8-r"

scaley = latticeh/100.0
matrix = psMat.scale(scaley, scaley)

newFont.layers[0].is_quadratic = True;
vecNames = []

for f in glob.glob(fontdir + "/*.png"):
    vecNames.append(f)

vecNames.sort()

for f in vecNames:

    cp = int(removePostfix(removePrefix(f, fontdir+"/"), ".png"))
    newFont.createMappedChar(cp)
    currGlyph = newFont[cp]
    currGlyph.importOutlines(f)
    # fontforge's autoTrace will invoke either autotrace or potrace
    # depending on what is installed on the system. In general,
    # potrace gives better results. This tracing step should probably
    # be moved into the c component, using libpotrace.
    currGlyph.autoTrace()
    currGlyph.width = latticew
    currGlyph.transform(matrix)
    currGlyph.simplify()

    # If fontforge sees a nearly blank character, it won't ouput it,
    # which will cause errors in the resulting pdf. Setting the width
    # manually should fix this, but this check is in here to make
    # sure.
    if (not currGlyph.isWorthOutputting()):
        print (f + " not worth outputting, failed to render character")
    

# Not sure about this part. Fontforge was complaining about invalid
# cvt and prep tables, during autoInstr in the first loop, so we just
# clear them, and autoInstr in a separate loop.
newFont.setTableData('cvt', None)
newFont.setTableData('prep', None)

for currGlyph in newFont.glyphs():  
    currGlyph.autoInstr()

fn = "SmoothScans" + str(fontnum)
newFont.fontname = fn
newFont.fullname = fn
newFont.familyname = fn
newFont.comment = "Generated by smoothscan"
# By default, fontforge includes the username in the copyright. We
# want to respect our user's privacy, so we clear it for them.
newFont.copyright = ""

newFont.generate(outname)



src/smoothscan.c
/*
  This file is part of smoothscan.

  smoothscan is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  smoothscan is distributed in the hope that it will be useful, but
  WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
  General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with smoothscan. If not, see <http://www.gnu.org/licenses/>.
*/

#define _GNU_SOURCE

/* Standard headers */
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <errno.h>

/* POSIX specific headers */
#include <unistd.h>
#include <sys/stat.h>
#include <sys/types.h>

#include <ftw.h>

/* GNU specific headers (Work on Windows?)*/
#include <getopt.h>

/* 3rd party library headers */
#include <leptonica/allheaders.h>
#include <hpdf.h>
/* #include <potracelib.h> */

#include "smoothscan.h"

int
main (int argc, char *argv[])
{
  struct args *args = parse_args (argc, argv);

  validate_args (args);

  JBDATA *data =
    classify_components (args->num_input_files, args->input_files,
			 args->thresh, args->weight);

  /* Render output of leptonica's classifier, if requested */
  if (args->debug_render_pages)
    {
      int i;
      PIXA *pa = jbDataRender (data, 0);
      for (i = 0; i < pa->n; i++)
	{
	  PIX *pix = pa->pix[i];
	  char filename[512];
	  sprintf (filename, "rendered_%05d.png", i);
	  pixWrite (filename, pix, IFF_PNG);
	}
    }

  struct mapping *maps = NULL;
  int num_fonts = register_mappings (data, &maps);

  char *tmpdirname = NULL;

  if (!args->debug_skip_font_gen)
    {
      tmpdirname = generate_fonts (data, maps, num_fonts, args->debug_tmpdir);
    }

  generate_pdf (args->outname, tmpdirname, num_fonts, args->num_input_files,
		data, maps, args->debug_draw_borders);

  /* clean up tmpdir */

  if (!args->debug_no_clean_tmpdir)
    {
      /* This may not be Windows compatible */
      if (nftw (tmpdirname, delete_file, 64, FTW_DEPTH | FTW_PHYS) == -1)
	{
	  error_quit ("Failed to clean up tmpdir.");
	}
    }

  if (args->debug_tmpdir == NULL)
    {
      free (tmpdirname);
    }

  free (maps);
  jbDataDestroy (&data);
  free (args->input_files);
  free (args);
  return 0;
}

unsigned char
first_code_point ()
{
  return 33;
}

unsigned char
max_code_point ()
{
  return 255;
}

unsigned char
next_code_point (unsigned char prev)
{
  /* 128 to 153 */
  if (prev == 126)
    return 128;
  /* 155 to 255 */
  if (prev == 153)
    return 155;

  return prev + 1;
}

void
print_help ()
{
  printf (
          "Usage: smoothscan [debug-options] [options] -o output.pdf inputs\n"
          "\n"
	  "Please read the man page for more in depth information.\n"
	  "inputs is the list of 1bpp TIFF files, one file per page\n"
          "\n"
	  "Regular Options:\n"
	  "    -o, --output FILE : Place the output into FILE.\n"
	  "    -t, --thresh VALUE\n"
	  "        Specify the threshold value [0.40 - 0.98], Default 0.85.\n"
	  "    -w, --weight VALUE\n"
	  "        Specify the weight value [0.0 - 1.0], Default 0.5.\n"
	  "    -h, --help\n"
	  "        Display basic usage information.\n"
	  "    -v, --version\n"
	  "        Display version information.\n"
          "\n"
	  "Debug options:\n"
	  "    --debug-tmpdir TMPDIR\n"
	  "        Use specified tmpdir instead of system tmpdir.\n"
	  "    --debug-draw-borders\n"
	  "        Draw red rectangles around each glyph in output pdf.\n"
	  "    --debug-render-pages\n"
	  "        Render output to image files in addition to pdf output.\n"
	  "    --debug-skip-font-gen\n"
	  "        Skip font generation step. Won't work if tmpdir doesn't already have fonts in it.\n"
	  "    --debug-no-clean-tmpdir\n"
	  "        Don't delete temporary files from tmpdir when processing is done.\n"
	  "\n"
	  "Report bugs to nate@natecraun.net or on the Github bug tracker\n"
	  "Smoothscan homepage: <https://natecraun.net/projects/smoothscan/>\n"
	  "Github Project page: <https://github.com/ncraun/smoothscan>\n"
          );
}

void
print_version ()
{
  printf ("Smoothscan " SMOOTHSCAN_VERSION "\n"
	  "Copyright (C) 2013 Nate Craun\n"
	  "License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>\n"
	  "This is free software: you are free to change and redistribute it.\n"
	  "There is NO WARRANTY, to the extent permitted by law.\n");

}

void
error_quit (const char *str)
{
  fprintf (stderr, "Error: %s\nSystem Error: %s\n", str, strerror (errno));
  exit (0);
}

int
num_digits (int a)
{
  int n = 0;

  if (a == 0)
    return 1;

  while (a > 0)
    {
      a /= 10;
      n++;
    }

  return n;
}

int
delete_file (const char *path, const struct stat *sb, int typeflag,
	     struct FTW *ftwbuf)
{
  int ret = remove (path);

  if (ret == -1)
    {
      error_quit ("Could not remove temp file.");
    }

  return ret;
}

void
pdf_error_handler (HPDF_STATUS error_no, HPDF_STATUS detail_no,
		   void *user_data)
{
  fprintf (stderr, "libharu error: error_no=%04X, detail_no=%d\n",
	   (unsigned int) error_no, (int) detail_no);
  error_quit ("PDF Generation problem.");
}


void *
malloc_guarded (size_t size)
{
  void *vptr = malloc (size);

  if (vptr == NULL)
    {
      error_quit ("Out of memory.");
    }

  return vptr;
}

int
file_exists (const char *filename)
{
  return (access (filename, R_OK) != -1);
}


void
create_font_from_dir (const char *dirname, const char *fontname, int latticeh,
		      int latticew, int fontnum)
{
  /* Call python from here */
  int commandlen =
    strlen ("smoothscan-fontgen.py") + 1 + strlen (dirname) + 1 +
    strlen (fontname) + 1 + num_digits (latticeh) + 1 +
    num_digits (latticew) + 1 + num_digits (fontnum);

  char *command = malloc_guarded (commandlen + 1);
  sprintf (command, "smoothscan-fontgen.py %s %s %d %d %d", dirname,
	   fontname, latticeh, latticew, fontnum);
  system (command);
  free (command);
}


char*
generate_fonts (const JBDATA * data, const struct mapping *maps,
		int num_fonts, char *dir)
{
  int dirnamelen = 0;
  char *dirname = NULL;

  if (dir == NULL)
    {

      /* Create a file in the system temp dir (usually /tmp) */
      /* This may not be portable to Windows */
      char suffix[] = "smoothscan_XXXXXX";
      /* Use the value of environmental var TMPDIR if it is set */
      char *tmpdir = getenv ("TMPDIR");
      if (tmpdir == NULL)
	{
	  /* Load the default temp directory */
	  tmpdir = P_tmpdir;
	}

      int tmpdirlen = strlen (tmpdir);
      int suffixlen = strlen (suffix);

      dirnamelen = tmpdirlen + suffixlen + 1;

      dirname = malloc_guarded (dirnamelen + 1);

      sprintf (dirname, "%s/%s", tmpdir, suffix);

      if (mkdtemp (dirname) == NULL)
	{
	  error_quit ("Failed to create main temp directory.");
	}
    }
  else
    {
      if (mkdir (dir, 0700) == -1)
	{
	  if (errno != EEXIST)
	    {
	      error_quit ("Couldn't make tmpdir.");
	    }
	}
      dirname = dir;
      dirnamelen = strlen (dir);
    }

  int i;

  PIXA *templates =
    pixaCreateFromPix (data->pix, data->nclass, data->latticew,
		       data->latticeh);

  if (templates == NULL)
    {
      error_quit ("Could not create templates from JBDATA.");
    }

  /* Create a temp dir for each font */
  char **fontdirnames = malloc_guarded (num_fonts * sizeof (char *));
  int *fontdirlens = malloc_guarded (num_fonts * sizeof (int *));

  for (i = 0; i < num_fonts; i++)
    {
      /* 1 for / 8 for %08d */
      fontdirlens[i] = dirnamelen + 1 + 8;
      fontdirnames[i] = malloc_guarded (fontdirlens[i] + 1);

      sprintf (fontdirnames[i], "%s/%08d", dirname, i);

      if (mkdir (fontdirnames[i], 0700) == -1)
	{
	  error_quit ("Failed to create font temp directory.");
	}
    }


  for (i = 0; i < templates->n; i++)
    {
      l_int32 iclass;
      l_int32 x;
      l_int32 y;

      numaGetIValue (data->naclass, i, &iclass);
      ptaGetIPt (data->ptaul, i, &x, &y);
      PIX *pix_clone = pixaGetPix (templates, i, L_CLONE);	/* the template */
      double wp1 = pixGetWidth (pix_clone);
      double hp1 = pixGetHeight (pix_clone);
      PIX *pix_padded =
	pixAddBorderGeneral (pix_clone, 0, data->latticew - wp1, 0,
			     data->latticeh - hp1, 0);

      if (pix_padded == NULL)
	{
	  error_quit ("Failed to add border to image.");
	}

      int fontn = maps[i].font_num;
      int code_point = maps[i].code_point;

      /* 1 for '/', 3 for %03d, 4 for '.png' */
      int filenamelen = fontdirlens[fontn] + 1 + 3 + 4;
      char *filename = malloc_guarded (filenamelen + 1);

      sprintf (filename, "%s/%03d.png", fontdirnames[fontn], code_point);

      if (pixWrite (filename, pix_padded, IFF_PNG) == 1)
	{
	  printf ("pixWrite failed to write %s.\n", filename);
	  error_quit ("Could not write to file.");
	}

      pixDestroy (&pix_clone);
      pixDestroy (&pix_padded);
      free (filename);
    }

  pixaDestroy (&templates);

  /* This part probably won't port over to Windows as well */

  /* TODO: parallelize this */
  for (i = 0; i < num_fonts; i++)
    {
      /* 1 for '/', 8 for %08d, 4 for '.ttf' */
      int fontnamelen = dirnamelen + 1 + 8 + 4;
      char *fontnamestr = malloc_guarded (fontnamelen + 1);
      sprintf (fontnamestr, "%s/%08d.ttf", dirname, i);

      create_font_from_dir (fontdirnames[i], fontnamestr, data->latticeh,
			    data->latticew, i);
      free (fontnamestr);
    }

  /* clean up */
  for (i = 0; i < num_fonts; i++)
    {
      free (fontdirnames[i]);
    }
  free (fontdirnames);
  free (fontdirlens);

  return dirname;
}

void
generate_pdf (const char *outname, const char *tmpdirname, int num_fonts,
	      int num_input_files, const JBDATA * data,
	      const struct mapping *maps, int debug_draw_borders)
{
  int i, j;
  /* Create the pdf document */
  l_int32 ncomp = numaGetCount (data->naclass);
  HPDF_Doc pdf = HPDF_New (pdf_error_handler, NULL);

  HPDF_SetCompressionMode (pdf, HPDF_COMP_ALL);
  HPDF_Font *fonts = malloc_guarded (num_fonts * (sizeof (HPDF_Font)));

  int dirlen = strlen (tmpdirname);

  /* Load the fonts */
  for (i = 0; i < num_fonts; i++)
    {
      /* 1 for '/', 8 for %08d, 4 for '.ttf' */
      int fontlen = dirlen + 1 + 8 + 4;
      char *font_tfname = malloc_guarded (fontlen + 1);
      sprintf (font_tfname, "%s/%08d.ttf", tmpdirname, i);

      const char *font_name =
	HPDF_LoadTTFontFromFile (pdf, font_tfname, HPDF_TRUE);
      fonts[i] = HPDF_GetFont (pdf, font_name, "KOI8-R");
      free (font_tfname);
    }

  for (j = 0; j < num_input_files; j++)
    {
      /* Add page to document */
      HPDF_Page pg = HPDF_AddPage (pdf);

      HPDF_Page_SetWidth (pg, data->w);
      HPDF_Page_SetHeight (pg, data->h);

      /* TODO: Boost efficiency by making startcomp go to the next page on the next iteration */
      int start_comp = 0;

      for (i = start_comp; i < ncomp; i++)
	{
	  l_int32 ipage;
	  l_int32 iclass;
	  l_int32 x;
	  l_int32 y;

	  numaGetIValue (data->napage, i, &ipage);

	  if (ipage != j)
	    continue;

	  numaGetIValue (data->naclass, i, &iclass);
	  ptaGetIPt (data->ptaul, i, &x, &y);

	  /*double left = x;
	     double top = data->h - y;
	     double right = x + data->latticew;
	     double bottom = data->h - (y + data->latticeh); */

	  char text[2];
	  text[0] = maps[iclass].code_point;
	  text[1] = '\0';

	  HPDF_Font font = fonts[maps[iclass].font_num];

	  HPDF_Page_BeginText (pg);
	  double fontsize = 100;

	  if (fontsize > 300)
	    {
	      error_quit ("This is a known bug.\n"
			  "libharu can't handle fontsizes bigger than 300, which is what is being requested here.\n"
			  "Please report this bug, and the file that produced this error");
	    }

	  HPDF_Page_SetFontAndSize (pg, font, 100);

	  HPDF_Page_MoveTextPos (pg, x, (data->h - y) - data->latticeh);
	  HPDF_Page_ShowText (pg, text);

	  HPDF_Page_EndText (pg);

	  if (debug_draw_borders)
	    {
	      HPDF_Page_SetRGBStroke (pg, 1, 0, 0);
	      /* In this, x, y is the LOWER LEFT, not UPPER LEFT */
	      HPDF_Page_Rectangle (pg, x, (data->h - y) - data->latticeh,
				   data->latticew, data->latticeh);
	      HPDF_Page_Stroke (pg);
	    }
	}
    }

  /* Output */
  HPDF_SaveToFile (pdf, outname);

  /* Cleanup */
  HPDF_Free (pdf);
  free (fonts);
}

JBDATA *
classify_components (int num_input_files, char **input_files, double thresh,
		     double weight)
{

  /* JBCLASSER* classer = jbCorrelationInit(JB_CONN_COMPS, 9999, 9999, thresh, weight); */
  JBCLASSER *classer =
    jbCorrelationInitWithoutComponents (JB_CONN_COMPS, 9999, 9999, thresh,
					weight);

  if (classer == NULL)
    {
      error_quit ("Unable to create leptonica JBCLASSER.");
    }

  int i;

  for (i = 0; i < num_input_files; i++)
    {
      PIX *page = pixRead (input_files[i]);

      if (page == NULL)
	{
	  printf ("Problem with page %s\n", input_files[i]);
	  error_quit ("Unable to read Page");
	}

      if (pixGetDepth (page) != 1)
	{
	  printf ("Input file %s is not 1bpp\n", input_files[i]);
	  error_quit
	    ("Only 1bpp (black and white) images currently supported.");
	}

      if (jbAddPage (classer, page) == 1)
	{
	  printf ("Problem with page %s\n", input_files[i]);
	  error_quit ("Unable to add page to JBCLASSIFIER.");
	}

      pixDestroy (&page);
    }

  /* This is the part we have to add stuff to save each glyph separately */
  JBDATA *data = jbDataSave (classer);

  if (data == NULL)
    {
      error_quit ("Unable to create the leptonica JBDATA.");
    }

  jbClasserDestroy (&classer);

  return data;
}

int
register_mappings (const JBDATA * data, struct mapping **in_maps)
{
  int i;
  /* Register mappings for each component */
  l_int32 ncomp = numaGetCount (data->naclass);

  *in_maps = malloc_guarded (ncomp * sizeof (struct mapping));
  struct mapping *maps = *in_maps;
  for (i = 0; i < ncomp; i++)
    {
      maps[i].used = 0;
    }

  unsigned char code_point = first_code_point ();
  int font_num = 0;

  for (i = 0; i < data->nclass; i++)
    {
      l_int32 iclass = i;

      if (maps[iclass].used)
	continue;

      maps[iclass].iclass = iclass;
      maps[iclass].font_num = font_num;
      maps[iclass].code_point = code_point;
      maps[iclass].used = 1;

      if (code_point == max_code_point ())
	{
	  code_point = first_code_point ();
	  font_num++;
	}
      else
	{
	  code_point = next_code_point (code_point);
	}
    }
  printf ("%d fonts\n", font_num + 1);

  return font_num + 1;
}

struct args *
parse_args (int argc, char *argv[])
{

  if (argc < 2)
    {
      print_help ();
      exit (0);
    }

  struct args *args = malloc_guarded (sizeof (struct args));

  args->num_input_files = 0;
  args->input_files = NULL;
  args->outname = NULL;
  args->thresh = .85;
  args->weight = .5;

  args->help_flag = 0;
  args->version_flag = 0;

  args->debug_tmpdir = NULL;
  args->debug_draw_borders = 0;
  args->debug_render_pages = 0;
  args->debug_skip_font_gen = 0;
  args->debug_no_clean_tmpdir = 0;

  /* Process Command Line args */
  int c;

  struct option long_options[] = {
    {"output", required_argument, 0, 'o'},
    {"help", no_argument, &args->help_flag, 1},
    {"version", no_argument, &args->version_flag, 1},
    {"thresh", required_argument, 0, 't'},
    {"weight", required_argument, 0, 'w'},

    /* Debug options */
    {"debug-tmpdir", required_argument, 0, 0},
    {"debug-draw-borders", no_argument, &args->debug_draw_borders, 1},
    {"debug-render-pages", no_argument, &args->debug_render_pages, 1},
    {"debug-skip-font-gen", no_argument, &args->debug_skip_font_gen, 1},
    {"debug-no-clean-tmpdir", no_argument, &args->debug_no_clean_tmpdir, 1},
    {0, 0, 0, 0}
  };

  while (1)
    {
      int option_index = 0;

      c = getopt_long (argc, argv, "o:t:w:", long_options, &option_index);

      if (c == -1)
	break;

      switch (c)
	{
	case 0:
	  {
	    if (strcmp ("debug-tmpdir", long_options[option_index].name) == 0)
	      {
		args->debug_tmpdir = optarg;
	      }
	    break;
	  }
	case 'o':
	  {
	    printf ("output: %s\n", optarg);
	    args->outname = optarg;
	    break;
	  }
	case 't':
	  {
	    double value;
	    sscanf (optarg, "%lf", &value);
	    args->thresh = value;
	    break;
	  }
	case 'w':
	  {
	    double value;
	    sscanf (optarg, "%lf", &value);
	    args->weight = value;
	    break;
	  }
	default:
	  print_help ();
	  abort ();
	}
    }

  if (args->help_flag)
    {
      print_help ();
      free (args);
      exit (EXIT_SUCCESS);
    }

  if (args->version_flag)
    {
      print_version ();
      free (args);
      exit (EXIT_SUCCESS);
    }


  if (optind < argc)
    {
      args->num_input_files = argc - optind;
      args->input_files =
	malloc_guarded ((args->num_input_files) * sizeof (char *));
      printf ("%d Input Files:\n", args->num_input_files);
      int k = 0;
      while (optind < argc)
	{
	  args->input_files[k] = argv[optind];
	  optind++;
	  k++;
	}
    }
  else
    {
      error_quit ("No input files specified.");
    }

  return args;
}

int
validate_args (const struct args *args)
{
  int i;
  if (args->num_input_files <= 0 || args->input_files == NULL)
    {
      error_quit ("No input files specified.");
    }
  if (args->outname == NULL)
    {
      error_quit ("No output file specified.");
    }
  /* Check that all input files exist */
  for (i = 0; i < args->num_input_files; i++)
    {
      if (!file_exists (args->input_files[i]))
	{
	  printf ("Can't read %s.\n", args->input_files[i]);
	  error_quit ("Input file doesn't exist.");
	}
    }
  /* 
     Check thresh and weight in valid range
     thresh (value for correlation score: in [0.4 - 0.98])
     weightfactor (corrects thresh for thick characters [0.0 - 1.0])
   */
  if (args->thresh < 0.4 || args->thresh > 0.98)
    {
      error_quit ("Threshold parameter must be in the range [0.4 - 0.98]");
    }
  if (args->weight < 0.0 || args->weight > 1.0)
    {
      error_quit ("Weight must be in range [0.0 - 1.0]");
    }
  /* Confirm overwriting if outname exists */
  if (file_exists (args->outname))
    {
      char c = 'n';
      printf ("Output file %s already exists. Overwrite? (y/N) ",
	      args->outname);
      scanf ("%c", &c);
      if (c != 'Y' && c != 'y')
	{
	  error_quit ("Output file exists.");
	}
    }

  return 0;
}
